curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,620 @@
+        Member, ProposeConfChangeRequest, ProposeId, ProposeRequest, Protocol, PublishRequest,
+        ReadState, ShutdownRequest, WaitSyncedRequest,
+    },
+};
+
+use super::{ClientApi, LeaderStateUpdate, ProposeResponse};
+
+/// Leader state of a client
+#[derive(Debug, Default)]
+struct LeaderState {
+    /// Current leader
+    leader: Option<ServerId>,
+    /// Current term
+    term: u64,
+}
+
+impl LeaderState {
+    /// Check the term and leader id, update the state if needed.
+    fn check_and_update(&mut self, leader_id: Option<u64>, term: u64) {
+        match self.term.cmp(&term) {
+            Ordering::Less => {
+                // reset term only when the resp has leader id to prevent:
+                // If a server loses contact with its leader, it will update its term for election. Since other servers are all right, the election will not succeed.
+                // But if the client learns about the new term and updates its term to it, it will never get the true leader.
+                if let Some(new_leader_id) = leader_id {
+                    debug!("client term updates to {term}");
+                    self.term = term;
+                    debug!("client leader id updates to {new_leader_id}");
+                    self.leader = Some(new_leader_id);
+                }
+            }
+            Ordering::Equal => {
+                if let Some(new_leader_id) = leader_id {
+                    if self.leader.is_none() {
+                        debug!("client leader id updates to {new_leader_id}");
+                        self.leader = Some(new_leader_id);
+                    }
+                    assert_eq!(
+                        self.leader,
+                        Some(new_leader_id),
+                        "there should never be two leader in one term"
+                    );
+                }
+            }
+            Ordering::Greater => {
+                debug!("ignore old term({term}) from server");
+            }
+        }
+    }
+}
+
+/// Cluster state of a client, thread safe
+struct ClusterState {
+    /// Cluster version
+    version: AtomicU64,
+    /// Cluster connects
+    connects: DashMap<ServerId, Arc<dyn ConnectApi>>,
+}
+
+impl Debug for ClusterState {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ClusterState")
+            .field("version", &self.version)
+            .field(
+                "connects",
+                &self.connects.iter().map(|v| *v.key()).collect_vec(),
+            )
+            .finish()
+    }
+}
+
+impl ClusterState {
+    /// Check the cluster version and update the connects if needed.
+    async fn check_and_update(
+        &self,
+        cluster_version: u64,
+        member_addrs: HashMap<ServerId, Vec<String>>,
+    ) -> Result<(), tonic::transport::Error> {
+        // Check and update the cluster version in client side
+        // FIXME: is these ordering options correct?
```

Comment:
It depends on how we use this version.  If what you need is an atomic variable, I think the Relaxed Ordering is enough.

Reply:
I don't get it. Can Relaxed Ordering here also serve as an optimistic lock?

Reply:
It it refactoed now